reframe/utility/jsonext.py

```diff
@@ -40,14 +41,35 @@ def encode(obj):
     return None
 
 
+def _make_json_friendly(obj):
+    '''Fallback type converter if json dump(s) raises a TypeError.'''
+
+    _valid_keys = {str, int, float, bool, type(None)}
+    if isinstance(obj, MutableMapping):
+        if not all((any(isinstance(k, t) for t in _valid_keys) for k in obj)):
+            newobj = type(obj)()
+            for k, v in obj.items():
+                newobj[util.toalphanum(str(k))] = v
```

Comment:
I'd prefer to use here the `encode()` function above, because it encodes correctly ReFrame types + some other builtin types and it promotes consistency between keys and values. The shortcoming is that `json.dump` does not treat consistently keys and values, so that the `encode()` will not be called for tuples if they are encountered as values, but we would call it here if tuples are encountered as keys. We could expand `encode()` to treat tuples as well, but I wouldn't go as far as to allow anything to be converted. That's why `encode()` simply returns `None` instead of `str(obj)` (n.b. it doesn't raise a `TypeError` also on purpose, in order to allow unknown objects to dumped as `"null"` instead of crashing). In any case, I think we should use `encode()` here for consistency with the values. Also, I wouldn't convert all the keys, but only those that are not in `_valid_keys`.

Reply:
Makes sense. I moved this into the `encode` function and all of a sudden, the test I had started failing again ... and the reason is be cause this patch was buggy! Say that we have the following tests:
```python
@rfm.simple_test
class MyTest(rfm.RunOnlyRegressionTest):
    ...
    custom_dict = variable(dict, value={
            ('a', 'b'): 'bananas',
            'a': 'bananas',
            ('a', 'c', 'b'): 'bananas',
    })
```
Here ReFrame will log the `custom_dict` variable independently and also will log the `MyTest` class (including its `self.__dict__`) which also contains the `custom_dict` variable. Before, this fix was only catching the `TypeError` when logging the `custom_dict`, and it would just return `None` when a `TypeError` was raised from attempting to log the `MyTest` class. Now that we have moved this functionality into the `encode` function, this function would "catch" the `MyTest` class after the `TypeError` and reattempt the logging without fixing the issues with the `self.__init__`, raising a `TypeError` (again).
The fix to this is to have a separate `encode_dict` function (which `encode` also uses) and this function will massage the `self.__dict__` from the `__rfm_json_encode__` function. This `encode_dict` can be forced to recur through potentially nested dictionaries, so it will also catch tuples as dict keys in sub-dictionaries, even when all the keys from the top-level dict are json-friendly (which was exactly the case of the `self.__dict__` from `MyTest`).

I'll push the changes and the unit tests in a sec.